File: vipaq/packages/binacle-vipaq/src/layouts/columnarCodec.ts
import {Coordinates, Dimensions, Header, Item} from "../models";
import {ProtocolReader} from "../ProtocolReader";
import {ProtocolWriter} from "../ProtocolWriter";

// Ports C#: ColumnarCodec. Each field for every item before the next field — six runs, each `count` values long
// (PROTOCOL.md §3.2):
//
//   L L L ... | W W W ... | H H H ... | X X X ... | Y Y Y ... | Z Z Z ...
//
// Like magnitudes sit next to each other, which usually compresses better. Uncompressed it is exactly as long
// as row-major.
export function writeColumnar(writer: ProtocolWriter, items: (Dimensions & Coordinates)[], header: Header): void {
	const dimensionsWidth = header.itemDimensionsWidth;
	const coordinatesWidth = header.itemCoordinatesWidth;

	for (const item of items) {
		writer.writeValue(item.length, dimensionsWidth);
	}

	for (const item of items) {
		writer.writeValue(item.width, dimensionsWidth);
	}
	
	for (const item of items){
		writer.writeValue(item.height, dimensionsWidth);
	}
	
	for (const item of items){
		writer.writeValue(item.x, coordinatesWidth);
	}
	
	for (const item of items){
		writer.writeValue(item.y, coordinatesWidth);
	}
	
	for (const item of items) {
		writer.writeValue(item.z, coordinatesWidth);
	}
}

export function readColumnar(reader: ProtocolReader, items: Item[], header: Header): void {
	const dimensionsWidth = header.itemDimensionsWidth;
	const coordinatesWidth = header.itemCoordinatesWidth;

	for (const item of items) {
		item.length = reader.readValue(dimensionsWidth);
	}
	
	for (const item of items) {
		item.width = reader.readValue(dimensionsWidth);
	}
	
	for (const item of items) {
		item.height = reader.readValue(dimensionsWidth);
	}
	
	for (const item of items){
		item.x = reader.readValue(coordinatesWidth);
	}
	
	for (const item of items) {
		item.y = reader.readValue(coordinatesWidth);
	}
	
	for (const item of items) {
		item.z = reader.readValue(coordinatesWidth);
	} 
}

File: vipaq/src/Binacle.ViPaq/Layouts/ColumnarCodec.cs
using System.Numerics;
using Binacle.Geometry;

namespace Binacle.ViPaq.Layouts;

// Each field for every item before the next field — six runs, each `count` values long (PROTOCOL.md §3.2):
//
//   L L L ... | W W W ... | H H H ... | X X X ... | Y Y Y ... | Z Z Z ...
//
// Like magnitudes sit next to each other, which usually compresses better. Uncompressed it is exactly as long
// as row-major. Whether it pays is unmeasured.
internal sealed class ColumnarCodec<T> : ILayoutEncoder<T>, ILayoutDecoder<T>
	where T : struct, IBinaryInteger<T>
{
	public void Write<TItem>(ProtocolWriter<T> protocolWriter, IReadOnlyList<TItem> items, Header header)
		where TItem : IWithReadOnlyDimensions<T>, IWithReadOnlyCoordinates<T>
	{
		var dimensionsWidth = header.ItemDimensionsWidth;
		var coordinatesWidth = header.ItemCoordinatesWidth;

		foreach (var item in items)
		{
			protocolWriter.WriteValue(item.Length, dimensionsWidth);
		}

		foreach (var item in items)
		{
			protocolWriter.WriteValue(item.Width, dimensionsWidth);
		}

		foreach (var item in items)
		{
			protocolWriter.WriteValue(item.Height, dimensionsWidth);
		}

		foreach (var item in items)
		{
			protocolWriter.WriteValue(item.X, coordinatesWidth);
		}

		foreach (var item in items)
		{
			protocolWriter.WriteValue(item.Y, coordinatesWidth);
		}

		foreach (var item in items)
		{
			protocolWriter.WriteValue(item.Z, coordinatesWidth);
		}
	}

	public void Read<TItem>(ProtocolReader<T> protocolReader, TItem[] items, Header header)
		where TItem : IWithDimensions<T>, IWithCoordinates<T>
	{
		var dimensionsWidth = header.ItemDimensionsWidth;
		var coordinatesWidth = header.ItemCoordinatesWidth;

		for (var index = 0; index < items.Length; index++)
		{
			items[index].Length = protocolReader.ReadValue(dimensionsWidth);
		}

		for (var index = 0; index < items.Length; index++)
		{
			items[index].Width = protocolReader.ReadValue(dimensionsWidth);
		}

		for (var index = 0; index < items.Length; index++)
		{
			items[index].Height = protocolReader.ReadValue(dimensionsWidth);
		}

		for (var index = 0; index < items.Length; index++)
		{
			items[index].X = protocolReader.ReadValue(coordinatesWidth);
		}

		for (var index = 0; index < items.Length; index++)
		{
			items[index].Y = protocolReader.ReadValue(coordinatesWidth);
		}

		for (var index = 0; index < items.Length; index++)
		{
			items[index].Z = protocolReader.ReadValue(coordinatesWidth);
		}
	}
}
